### ISSUE TITLE
My Drive Support & Options pattern

### ISSUE THREAD

**Issue by m-rots** [m-rots/bernard#7]
To truly be the go-to solution for caching Google Drive metadata, Bernard should support the personal My Drive space as well.

However, this will require multiple internal changes, and most importantly: breaking API changes.

### Internal changes
- allContent, changedContent, drive and pageToken (entire fetch stack) will need to support the new My Drive target by selectively disabling or changing query parameters (corpora, driveId and includeItemsFromAllDrives).
- The actual ID of the My Drive drive must be fetched on Bernard initialisation or must be stored in the datastore somehow. This is required to be inline with the new datastore tables of #5. The Drive ID can be fetched with a GET request for fileId "root". We should seek solutions allowing for some kind of persistence of the My Drive driveID, maybe by email address (this will require a new table in the datastore).

### Breaking API changes
- Bernard initialisation will radically change, as it will now be configured with "options". Options will not only allow the dependant to easily configure and swap between My Drive and Shared Drives, but also allow for future functionality without breaking the API.
Examples of the options pattern:
  - [Introduction to the options pattern](https://www.sohamkamani.com/golang/options-pattern/)
  - [Original idea by Rob Pike](https://commandcenter.blogspot.com/2014/01/self-referential-functions-and-design.html)
  - [Comparing solutions](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis)

### Unknown behaviour
- driveId is not provided in My Drive-targeted responses. Research must be done to see what happens if you move a file from My Drive to a Shared Drive, does the driveId show up? If so, for the fetching layer, the driveId must be empty for the comparison. If a driveId does show up, mark it as a removed file.

**Comment by m-rots** [comment]
## Proposal for the new API

To support multiple Drives to be used from one Bernard instance, without having to re-configure the options every time, I propose to change the PartialSync and FullSync signatures.

### FullSync and PartialSync

Bernard instances will now be coupled only to the authenticator and datastore, removing the DriveID from the equation. This change will allow one Bernard instance to track both the My Drive space of the authenticated user, as well as all the Shared Drives the user has access to.

Instead of passing the DriveID to Bernard itself, its functions PartialSync and FullSync will now require one to choose between My Drive and a Shared Drive.

#### Required options

```go
type syncOptions struct {
	driveID string
	myDrive bool
}

type SyncOption func(*syncOptions)

func WithMyDrive() SyncOption {
	return func(opts *syncOptions) {
		opts.myDrive = true
	}
}

func WithSharedDrive(driveID string) SyncOption {
	return func(opts *syncOptions) {
		opts.myDrive = false
		opts.driveID = driveID
	}
}
```

#### New function signatures

The FullSync function will now accept either the MyDrive or SharedDrive through the SyncOption.
Passing this parameter is required.
```go
func (bernard *Bernard) FullSync(opt SyncOption) error
```

The PartialSync function will now accept either the MyDrive or SharedDrive through the Sync Option.
Passing this parameter is required. However, additional hooks can still be supplied.
```go
func (bernard *Bernard) PartialSync(opt SyncOption, hooks ...Hook) error
```

### Datastore

The `Drive` column in the datastore currently exists of both an `id` and `pageToken` column.
I propose to add a third column: `owner`.

The owner column will accept both a string and a NULL value. If the value is NULL, then the Drive is a Shared Drive, as these have no single owner (highest up is the manager role). If the Drive is of the My Drive space, then the column will have the value of owner's email address, which should equal the email address as supplied through the Authenticator.

Currently the PartialSync function already fetches the pageToken from the store. Instead of fetching the drive's pageToken by ID, it will now fetch it by the owner column. In addition, the DriveID value will be retrieved as well to pass to store.PartialSync.

### Authenticator

In addition to the AccessToken function, the Authenticator will now also require the presence of an EmailAddress function which should retrieve the email address of the authenticated user. This function will be used inside the bernard.PartialSync function.

**Comment by m-rots** [comment]
## My Drive support

After initial testing, My Drive support turns out to be quite a pain...

### Problems
1. `Shared with me` files and folders cannot be filtered out API-wise.
2. `'me' in owners` causes Google Photos to appear for some reason, even with the spaces set to drive only.
3. The API endpoint `/files` is VERY slow, taking between 4 and 10 seconds per 1000 files (in contrast to 200-500ms).
4. Instead of `/files`, `/changes` could be used with `1` as the initial pageToken. However, this fetches all the previous "removed" items. ALL of them. Taking a considerable amount of time to fully fetch all of them.

### Moving forward

Unfortunately, My Drive will not be supported because of these problems. The extra time it takes to initially fetch all files and folders is beyond acceptable. Bernard will only support Shared Drives, as it looks like Google uses a different backend for these drives, offering much lower latency and less complexity overhead.

Options will still be part of a new API coming in version 0.3.0. These options will allow fine-grained control over some of Bernard's internal behaviour, such as:

- Sleeping N number of seconds (time.Duration) between the fetch of the startPageToken and the full sync. This will ensure no data is lost when changes are actively being made to the Shared Drive while a full sync is in progress.
- Custom JSON implementation.
- Custom HTTP client.
- Injection of custom rate limiting so requests are only made N times a second.

### TLDR
- My Drive will NOT be supported.
- Options are coming with exciting possibilities.
